aesd-char-driver/main.c

```diff
@@ -168,12 +180,136 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+    unsigned int offset_counter = 0;
+    long status = 0;
+    int i;
+
+
+    // check for valid cmd number
+    if(cmd >= AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED) {
+        PDEBUG("Command number too large\n");
+        status = -EINVAL;
+        goto leave_func;
+    }
+
+    // acquire mutex to access circular buffer 
+    if (mutex_lock_interruptible(&aesd_device.file_lock) != 0) {
+
+		PDEBUG("Mutex fail\n");
+		status = -ERESTARTSYS;
+        goto leave_func;
+	}
+
+    // check for valid offset 
+    if(device->circular_buffer.entry[cmd].size < cmd_offset) {
+        PDEBUG("cmd offset fail\n");
+        status = -EINVAL;
+        goto unlock_leave;
+    }
+
+    for(i = 0; i < cmd ; i++) {
+        // if null, error or do not account for size?
+        if(device->circular_buffer.entry[i].buffptr == NULL) {
+            continue;
+        }
+        offset_counter += device->circular_buffer.entry[i].size;
+    }
+
+    // once outside add offset 
+    offset_counter += cmd_offset;
+
+    // update fpos 
+    filp->f_pos = offset_counter;
+
+    unlock_leave:
+    // mutex unlock 
+    mutex_unlock(&aesd_device.file_lock);
+
+    leave_func: 
+    return status;
+}
+
+// based on scull ioctl design 
+long aesd_ioctl_unlocked (struct file *filp, unsigned int cmd, unsigned long arg) 
+{
+    long status = 0;
+    struct aesd_seekto copy_user_seek;
+    
+    //check for valid filp 
+    if(filp == NULL) {
+        PDEBUG("filp is null\n");
+        status = -EINVAL;
+        goto ioctl_leave;
+    }
+    
+    /*
+        * from text
+        * extract the type and number bitfields, and don't decode
+        * wrong cmds: return ENOTTY (inappropriate ioctl)
+    */
+
+    if (_IOC_TYPE(cmd) != AESD_IOC_MAGIC) {
+        PDEBUG("AESD_IOC_MAGIC fail\n");
+        status = -ENOTTY;
+        goto ioctl_leave;
+    } 
+
+    if (_IOC_NR(cmd) > AESDCHAR_IOC_MAXNR) {
+        PDEBUG("AESDCHAR_IOC_MAXNR fail\n");
+        status = -ENOTTY;
+        goto ioctl_leave;
+    }
+
+    if(cmd != AESDCHAR_IOCSEEKTO) {
+        PDEBUG("ioctl func cmd fail\n");
+        status = -EINVAL;
+        goto ioctl_leave;
+    }
+
+    switch(cmd) {
+
+       case AESDCHAR_IOCSEEKTO:  
+       {
+            // copy from user 
+            if (0 != (copy_from_user(&copy_user_seek, (const void __user*)arg, sizeof(struct aesd_seekto)))) {
+                status = -EFAULT;
+                goto ioctl_leave;
+            } else { 
+                status = aesd_adjust_file_offset(filp, copy_user_seek.write_cmd, copy_user_seek.write_cmd_offset);
+            }
+       } 
+       break;
+    // default case not required due to check for valid commmand above 
+    }
+
+    ioctl_leave: 
+
+    return status;
+}
+
+loff_t aesd_llseek(struct file *filp, loff_t off, int whence) 
+{
+    int status = 0;
+    // total size is updated after every add entry
+    status = fixed_size_llseek(filp, off, whence, total_buffer_size);
```

Comment:
@sachininja rather than using a global `total_buffer_size` here I'd use the actual value stored in the circular buffer, or re-calculate it.